src/miner.erl

```diff
@@ -600,9 +445,232 @@ terminate(Reason, _State) ->
+    {ok, ConsensusMembers} =
+        blockchain_ledger_v1:consensus_members(blockchain:ledger(Chain)),
+    length(ConsensusMembers).
+
+-spec hash_check_if_stale(H, [H], C) -> {ok, {}} | {error, E} when
+    E ::  {stale, H, C}
+        | {unexpected_counts, [{H, C}]},
+    H :: blockchain_block:hash(),
+    C :: non_neg_integer().
+hash_check_if_stale(HashCurr, Hashes, VotesNeeded) ->
+    case
+        lists:filter(
+            fun ({_, Votes}) -> Votes >= VotesNeeded end,
+            maps:to_list(miner_util:list_count(Hashes))
+        )
+    of
+        %% We expect every stamp to contain the same block hash:
+        [{HashCurr, _}]      -> {ok, {}};
+        [{HashStale, Votes}] -> {error, {stale, HashStale, Votes}};
+        Counts               -> {error, {unexpected_counts, Counts}}
+    end.
+
+-spec create_block(
+    metadata(),
+    blockchain_txn:txns(),
+    non_neg_integer(),
+    blockchain:blockchain(),
+    non_neg_integer(),
+    swarm_keys()
+) ->
+    create_block_ok().
+create_block(Metadata, Txns, HBBFTRound, Chain, VotesNeeded, {MyPubKey, SignFun}) ->
+    {ok, CurrentBlock} = blockchain:head_block(Chain),
+    HeightCurr = blockchain_block:height(CurrentBlock),
+    HeightNext = HeightCurr + 1,
+    Ledger = blockchain:ledger(Chain),
+    SnapshotHash = snapshot_hash(Ledger, HeightNext, Metadata, VotesNeeded),
+    SeenBBAs = metadata_to_seen_bbas(Metadata),
+    {ok, CurrentBlockHash} = blockchain:head_hash(Chain),
+    {SeenVectors, BBAs} = lists:unzip(SeenBBAs),
+    BBA = pick_consented_or_default(VotesNeeded, BBAs, <<>>),
+    {ElectionEpoch, EpochStart, TxnsToInsert, InvalidTransactions} =
+        select_transactions(Chain, Txns, CurrentBlock, HeightCurr, HeightNext),
+    NewBlock =
+        blockchain_block_v1:new(#{
+            prev_hash       =>  CurrentBlockHash,
+            height          =>  HeightNext,
+            transactions    =>  TxnsToInsert,
+            signatures      =>  [],
+            hbbft_round     =>  HBBFTRound,
+            time            =>  block_time(CurrentBlock,  Metadata),
+            election_epoch  =>  ElectionEpoch,
+            epoch_start     =>  EpochStart,
+            seen_votes      =>  SeenVectors,
+            bba_completion  =>  BBA,
+            snapshot_hash   =>  SnapshotHash
+        }),
+    BinNewBlock = blockchain_block:serialize(NewBlock),
+    Signature = SignFun(BinNewBlock),
+    lager:debug("Worker:~p, Created Block: ~p, Txns: ~p",
+                [self(), NewBlock, TxnsToInsert]),
+    #{
+        address               => libp2p_crypto:pubkey_to_bin(MyPubKey),
+        unsigned_binary_block => BinNewBlock,
+        signature             => Signature,
+
+        %% Both pending and invalid are to be removed from the buffer:
+        pending_txns          => TxnsToInsert,
+        invalid_txns          => InvalidTransactions
+    }.
+
+-spec block_time(blockchain_block:block(), metadata()) -> pos_integer().
+block_time(Block, Metadata) ->
```

Comment:
`block_time` -> `median_metadata_timestamp`

Reply:
That doesn't seem quite right to me: `block_time` describes WHAT the function computes, whereas `median_metadata_timestamp` describes a subset of of HOW it computes it.

Perhaps `new_block_time` or `next_block_time` or something like that? (If you object because `block_time` sounds like an accessor or something...)

Reply:
next block time is fine